app/controllers/CheckYourAnswersController.scala

```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
-            val errorList = errors.toChain.toList
-            val errorMessages = errorList.map(_.errorMessage).mkString("\n")
-            logger.error(s"Unable to create a VAT return request from user answers: $errorMessages")
-
-            Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
-        }
+
+      val redirectToFirstError = getRedirect(validate(period), period).headOption
+
+      (redirectToFirstError, incompletePromptShown) match {
+        case (Some(c), true) => Future.successful(Redirect(c))
+        case (Some(_), false) => Future.successful(Redirect(routes.CheckYourAnswersController.onPageLoad(period)))
+        case _ =>
+          val validatedVatReturnRequest =
+            vatReturnService.fromUserAnswers(request.userAnswers, request.vrn, period, request.registration)
+
+          val validatedCorrectionRequest = request.userAnswers.get(CorrectPreviousReturnPage).map(_ =>
+            correctionService.fromUserAnswers(request.userAnswers, request.vrn, period, request.registration.commencementDate))
+
+          (validatedVatReturnRequest, validatedCorrectionRequest) match {
+            case (Valid(vatReturnRequest), Some(Valid(correctionRequest))) =>
+              submitReturn(vatReturnRequest, Option(correctionRequest), period)
+            case (Valid(vatReturnRequest), None) =>
+              submitReturn(vatReturnRequest, None, period)
+            case (Invalid(vatReturnErrors), Some(Invalid(correctionErrors))) =>
+              val errors = vatReturnErrors ++ correctionErrors
+              val errorList = errors.toChain.toList
+              val errorMessages = errorList.map(_.errorMessage).mkString("\n")
+              logger.error(s"Unable to create a vat return and correction request from user answers: $errorMessages")
+              Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
+            case (Invalid(errors), _) =>
+              val errorList = errors.toChain.toList
+              val errorMessages = errorList.map(_.errorMessage).mkString("\n")
+              logger.error(s"Unable to create a vat return request from user answers: $errorMessages")
+              Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
+            case (_, Some(Invalid(errors))) =>
+              val errorList = errors.toChain.toList
+              val errorMessages = errorList.map(_.errorMessage).mkString("\n")
+              logger.error(s"Unable to create a Corrections request from user answers: $errorMessages")
+              Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
+          }
       }
+  }
+
+  //noinspection ScalaStyle
+  def getRedirect(errors: List[ValidationError], period: Period): List[Call] = {
+    errors.flatMap {
+      case DataMissingError(AllSalesFromNiQuery) =>
+        logger.error(s"Data missing - no data provided for NI sales")
+        Some(routes.CountryOfConsumptionFromNiController.onPageLoad(CheckMode, period, Index(0)))
```

Comment:
Is is right that this is always Index(0)?

Reply:
Yes, this is for when a user has answered "Yes" to NI sales, but have provided no further data.